### PR TITLE
feat(api): Build fail if new API doesn't have owner

### PR DESCRIPTION
**Pull request by sentaur-athena**
based on https://github.com/getsentry/sentry/pull/54876

Highlevel goal:
* Being able to track API ownership and improve APIs
* Implementing suggested solution here: https://www.notion.so/sentry/Tracking-API-Ownership-And-Publish-State-57e34e8defb4401fb081fdf257d6dbe5

This PR:
* Add known ownerships from CODEOWNERs
* Create an allowlist to prevent more unowned apis
* Write a test that would disallow creating unowned apis

Testing the test:
Removed ownership from one endpoint and got this error:

```
--> Building drf-spectacular openapi spec (combines with deprecated docs)
Traceback (most recent call last):
  File "/Users/athena1/code/sentry/.venv/bin/sentry", line 33, in <module>
    sys.exit(load_entry_point('sentry', 'console_scripts', 'sentry')())
  File "/Users/athena1/code/sentry/src/sentry/runner/__init__.py", line 193, in main
    func(**kwargs)
  File "/Users/athena1/code/sentry/.venv/lib/python3.8/site-packages/click/core.py", line 1128, in __call__
    return self.main(*args, **kwargs)
  File "/Users/athena1/code/sentry/.venv/lib/python3.8/site-packages/click/core.py", line 1053, in main
    rv = self.invoke(ctx)
  File "/Users/athena1/code/sentry/.venv/lib/python3.8/site-packages/click/core.py", line 1659, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/athena1/code/sentry/.venv/lib/python3.8/site-packages/click/core.py", line 1395, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/athena1/code/sentry/.venv/lib/python3.8/site-packages/click/core.py", line 754, in invoke
    return __callback(*args, **kwargs)
  File "/Users/athena1/code/sentry/.venv/lib/python3.8/site-packages/click/decorators.py", line 26, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/Users/athena1/code/sentry/src/sentry/runner/decorators.py", line 29, in inner
    return ctx.invoke(f, *args, **kwargs)
  File "/Users/athena1/code/sentry/.venv/lib/python3.8/site-packages/click/core.py", line 754, in invoke
    return __callback(*args, **kwargs)
  File "/Users/athena1/code/sentry/.venv/lib/python3.8/site-packages/click/decorators.py", line 26, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/Users/athena1/code/sentry/src/sentry/runner/commands/django.py", line 14, in django
    execute_from_command_line(argv=[ctx.command_path] + list(management_args))
  File "/Users/athena1/code/sentry/.venv/lib/python3.8/site-packages/django/core/management/__init__.py", line 419, in execute_from_command_line
    utility.execute()
  File "/Users/athena1/code/sentry/.venv/lib/python3.8/site-packages/django/core/management/__init__.py", line 413, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/Users/athena1/code/sentry/.venv/lib/python3.8/site-packages/django/core/management/base.py", line 354, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/Users/athena1/code/sentry/.venv/lib/python3.8/site-packages/django/core/management/base.py", line 398, in execute
    output = self.handle(*args, **options)
  File "/Users/athena1/code/sentry/.venv/lib/python3.8/site-packages/drf_spectacular/management/commands/spectacular.py", line 72, in handle
    schema = generator.get_schema(request=None, public=True)
  File "/Users/athena1/code/sentry/.venv/lib/python3.8/site-packages/drf_spectacular/generators.py", line 268, in get_schema
    paths=self.parse(request, public),
  File "/Users/athena1/code/sentry/.venv/lib/python3.8/site-packages/drf_spectacular/generators.py", line 188, in parse
    self._initialise_endpoints()
  File "/Users/athena1/code/sentry/.venv/lib/python3.8/site-packages/drf_spectacular/generators.py", line 171, in _initialise_endpoints
    self.endpoints = self.inspector.get_api_endpoints()
  File "/Users/athena1/code/sentry/.venv/lib/python3.8/site-packages/drf_spectacular/generators.py", line 26, in get_api_endpoints
    api_endpoints = hook(endpoints=api_endpoints)
  File "/Users/athena1/code/sentry/src/sentry/apidocs/hooks.py", line 80, in custom_preprocessing_hook
    raise SentryApiBuildError(
sentry.apidocs.utils.SentryApiBuildError: Please add owner attribute to your endpoint. If you can't find your team in ApiOwners feel free to add the associated github group. 
See https://develop.sentry.dev/api/public/#how-to-make-an-endpoint-public for more information.

```

### CODE REVIEW

File: src/sentry/apidocs/api_ownership_allowlist_dont_modify.py
Comment (asottile-sentry) [inline src/sentry/apidocs/api_ownership_allowlist_dont_modify.py:4]: you'll want to add yourself (or a team) to CODEOWNERS for this file -- otherwise it will regress unfortunately :/

**Review by asottile-sentry — COMMENTED** [review]
looks good!

**Review by EricHasegawa — COMMENTED** [review]
Looks great :) Will be interesting tracking down the allowlist endpoints after this

File: .github/CODEOWNERS
Comment (sentaur-athena) [reply .github/CODEOWNERS:415]: yes, it doesn't exist.

File: src/sentry/api/api_owners.py
Comment (sentaur-athena) [reply src/sentry/api/api_owners.py:21]: Doesn't matter really.

File: src/sentry/apidocs/api_ownership_allowlist_dont_modify.py
Comment (sentaur-athena) [reply src/sentry/apidocs/api_ownership_allowlist_dont_modify.py:4]: Yeah, I add this file to our github review so it requires review from us. Something like this: https://github.com/orgs/community/discussions/49633 
I don't have the right permissions so yeah, in progress.

**Review by JoshFerge — APPROVED** [review]
looks great! i'm also wondering if we should add a test that confirms that each endpoint is owned in CODEOWNERs as well, and that it matches to what they've set on the endpoint?

File: src/sentry/apidocs/api_ownership_allowlist_dont_modify.py
Comment (sentaur-athena) [reply src/sentry/apidocs/api_ownership_allowlist_dont_modify.py:4]: The file is already added to codeowners. Seems like we can't do above in our repo so I'll look into other ways.

**Comment by sentaur-athena** [comment]
@JoshFerge I added updating codeowners from apiowners to my stretch goals. I'm thinking they don't have to cause a failure but since we're pinging people with their apis in apiowners that would be more accurate source than codeowners moving forward.

**Comment by JoshFerge** [comment]
> @JoshFerge I added updating codeowners from apiowners to my stretch goals. I'm thinking they don't have to cause a failure but since we're pinging people with their apis in apiowners that would be more accurate source than codeowners moving forward.

no worries. yeah it may not even have to auto update. a test failure if the CODEOWNERs doesn't match with the APIOwner would be nice though

**Review by JoshFerge — APPROVED** [review]

**Pull request merged**
